### PR TITLE
Proposal for improving MatchPattern

### PR DESCRIPTION
**Pull request by grolu**
**What this PR does / why we need it**:
Refactored function and added tests

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```feature user

```

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -164,30 +164,44 @@ const (
...
+		if match != nil {
+			match.Garden = g.Name
+			matches = append(matches, match)
 		}
 	}
 
-	if patternMatch != nil {
-		// Did not match pattern of current garden, but did match other pattern
-		return patternMatch, nil
+	if len(matches) == 1 {
```
Comment (holgerkoser) [inline pkg/config/config.go:199]: We could early exit from loop and I would check first for errors and return successfully at the end.  
```go
        var patternMatch *PatternMatch
        
	for _, g := range config.Gardens {
		match, err := matchPattern(g.Patterns, value)
		if err != nil {
			return nil, err
		}

		if match != nil {
			match.Garden = g.Name
                         if patternMatch != nil {
                         	return nil, errors.New("the provided value resulted in an ambiguous match")
                         }
                         
                         patternMatch = match
		}
	}
	
	if patternMatch == nil {
	         return nil, errors.New("the provided value does not match any pattern")
	}
	
	return matches[0], nil
```

**Review by petersutter — APPROVED** [review]
/lgtm 

However I also like holgers suggestion, even more simplified

File: pkg/config/config_test.go [hunk 2]
```diff
@@ -31,20 +31,87 @@ var _ = Describe("Config", func() {
...
+		Entry(
+			"should succeed if pattern for value is (potential) non unique pattern and clusterIdentity1 is set in current target (non unique case / keep garden)",
+			clusterIdentity1,
+			"",
+			clusterIdentity1),
+		Entry(
+			"should succeed if pattern for value is (potential) non unique pattern and clusterIdentity2 is set in current target (non unique case / keep garden)",
+			clusterIdentity2,
+			"",
+			clusterIdentity2),
```
Comment (petersutter) [inline pkg/config/config_test.go:88]: is there a difference between these two? or could you get rid of one of them?

File: pkg/config/config_test.go [hunk 3]
```diff
@@ -31,20 +31,87 @@ var _ = Describe("Config", func() {
...
+	var patternValue = func(prefix string) string {
+		value := fmt.Sprintf("shoot--%s--%s", project, shoot)
+
+		if prefix != "" {
+			value = fmt.Sprintf("%s/%s", prefix, value)
+		}
+		return value
+	}
+
+	DescribeTable("Validation of valid target values",
```
Comment (holgerkoser) [inline pkg/config/config_test.go:56]:
```suggestion
	DescribeTable("MatchPattern returns a match",
```

File: pkg/config/config_test.go [hunk 4]
```diff
@@ -31,20 +31,87 @@ var _ = Describe("Config", func() {
...
+		func(currentGardenName string, patternPrefix string, expectedGarden string) {
+			value := patternValue(patternPrefix)
+			expectedPM := &config.PatternMatch{Garden: expectedGarden, Project: project, Shoot: shoot}
+			match, err := cfg.MatchPattern(currentGardenName, value)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(match).To(Equal(expectedPM))
+		},
+		Entry(
+			"should succeed if pattern for value is (potential) non unique pattern and garden is set in current target (unique case / keep garden)",
```
Comment (holgerkoser) [inline pkg/config/config_test.go:65]:
```suggestion
			"when the value contains a gardenName it's value is correctly extracted",
```

File: pkg/config/config_test.go [hunk 5]
```diff
@@ -31,20 +31,87 @@ var _ = Describe("Config", func() {
...
+			clusterIdentity2,
+			clusterIdentity1,
+			clusterIdentity1),
+		Entry(
+			"should succeed if pattern for value is (potential) non unique pattern and clusterIdentity1 is set in current target (non unique case / keep garden)",
+			clusterIdentity1,
+			"",
+			clusterIdentity1),
+		Entry(
+			"should succeed if pattern for value is (potential) non unique pattern and clusterIdentity2 is set in current target (non unique case / keep garden)",
```
Comment (holgerkoser) [inline pkg/config/config_test.go:85]:
```suggestion
			"when the targetValue does not contain a gardenName the given gardenName is taken",
```

File: pkg/config/config_test.go [hunk 6]
```diff
@@ -31,20 +31,87 @@ var _ = Describe("Config", func() {
...
+			"",
+			clusterIdentity1,
+			clusterIdentity1),
+		Entry(
+			"should succeed if pattern for value is (potential) non unique pattern and other garden is set in current target (unique case / switch garden)",
+			clusterIdentity2,
+			clusterIdentity1,
+			clusterIdentity1),
+		Entry(
+			"should succeed if pattern for value is (potential) non unique pattern and clusterIdentity1 is set in current target (non unique case / keep garden)",
```
Comment (holgerkoser) [inline pkg/config/config_test.go:80]: What is the difference between the last two tests?

File: pkg/config/config_test.go (hunk 4 above)
Comment (holgerkoser) [reply pkg/config/config_test.go:65]: Whats ist the difference between the first three tests?

File: pkg/config/config_test.go [hunk 7]
```diff
@@ -31,20 +31,87 @@ var _ = Describe("Config", func() {
...
+
+	DescribeTable("Validation of invalid target values",
+		func(currentGardenName string, patternPrefix string, expectedErrorString string) {
+			value := patternValue(patternPrefix)
+			_, err := cfg.MatchPattern(currentGardenName, value)
+			Expect(err).To(HaveOccurred())
+			Expect(err).To(MatchError(ContainSubstring(expectedErrorString)))
+		},
+		Entry(
+			"should fail if pattern for value is (potential) non unique pattern and garden is not set in current target (non unique case / ambiguous match / fail to determine garden)",
```
Comment (holgerkoser) [inline pkg/config/config_test.go:99]:
```suggestion
			"when no garden is given and the value matches multiple gardens",
```

File: pkg/config/config_test.go [hunk 8]
```diff
@@ -31,20 +31,87 @@ var _ = Describe("Config", func() {
...
+			"",
+			"",
+			"the provided value resulted in an ambiguous match"),
+		Entry(
+			"should fail if pattern for value is not found and garden is set",
```
Comment (holgerkoser) [inline pkg/config/config_test.go:104]:
```suggestion
			"when a garden is given and the value does not match any pattern",
```

File: pkg/config/config_test.go (hunk 2 above)
Comment (grolu) [reply pkg/config/config_test.go:88]: It is important to not change the cluster. I wanted to cover this by testing with both clusters. Depending on the code, one of those two tests could fail but it I cannot for-see which one (as it depends on the code)

File: pkg/config/config_test.go [hunk 9]
```diff
@@ -31,20 +32,92 @@ var _ = Describe("Config", func() {
...
+			clusterIdentity2,
+			"",
+			clusterIdentity2),
+	)
+
+	DescribeTable("MatchPattern returns an error",
+		func(currentGardenName string, patternPrefix string, expectedErrorString string) {
+			value := patternValue(patternPrefix)
+			_, err := cfg.MatchPattern(currentGardenName, value)
+			Expect(err).To(HaveOccurred())
```
Comment (holgerkoser) [inline pkg/config/config_test.go:96]: MatchError already ensure that an error occurs.
```suggestion
```

**Review by petersutter — APPROVED** [review]
/lgtm

**Pull request merged**